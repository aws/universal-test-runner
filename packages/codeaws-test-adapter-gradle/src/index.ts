// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import runCommand from './runCommand'
import log from './log'

import { AdapterInput, AdapterOutput } from '@sentinel-internal/codeaws-test-runner'

export async function executeTests({ testsToRun = [] }: AdapterInput): Promise<AdapterOutput> {
  const executable = 'gradle'
  const args = ['test']
  const testNamesToRun = testsToRun.map(({ testName }) => testName)
  if (testNamesToRun.length > 0) {
    args.push(...testNamesToRun.flatMap((testName) => ['--tests', `*${testName}`]))
  }
  log.stderr(`Running tests with gradle using command: ${[executable, ...args].join(' ')}`)
  const { status } = await runCommand(executable, args)
  return { exitCode: status ?? 1 }
}
