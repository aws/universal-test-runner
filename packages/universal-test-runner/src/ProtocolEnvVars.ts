// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export const ProtocolEnvVars = {
  /*
   * The version of the protocol being used, e.g. '0.1.0'
   */
  VERSION: 'TEP_VERSION',

  /*
   * Pipe-separated list of tests to be run. Suite name and filepath can be
   * specified optionally, using hashes. e.g.
   *
   * 'test1|test4|test9'
   * 'suite1#test1|suite2#test4|suite#test9'
   * 'file1.js#suite1#test1|file2.js#suite2#test4|file3.js#suite#test9'
   */
  TESTS_TO_RUN: 'TEP_TESTS_TO_RUN',

  TESTS_TO_RUN_FILE: 'TEP_TESTS_TO_RUN_FILE',

  LOG_FILE_NAME: 'TEP_LOG_FILE_NAME',

  REPORT_FORMAT: 'TEP_REPORT_FORMAT',
} as const
