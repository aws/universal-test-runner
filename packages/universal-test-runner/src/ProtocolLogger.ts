// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import fileSystem from 'fs/promises'
import path from 'path'
import { log } from './log'

type LogEntryType =
  | 'MESSAGE'
  | 'CUSTOM'
  | 'PROTOCOL_READ_START'
  | 'PROTOCOL_READ_END'
  | 'DISCOVERED_PROTOCOL_ENV_VARS'
  | 'TEST_RUN_START'
  | 'TEST_RUN_END'
  | 'PROTOCOL_VERSION'

type CustomLogEntryType = 'ADAPTER_LOAD_START' | 'ADAPTER_LOAD_END' | 'ADAPTER_PATH'

type LogLevel = 'INFO' | 'ERROR' | 'DEBUG' | 'WARN'

interface LogEntry {
  timestamp: number
  type: LogEntryType
  level: LogLevel
  data: any
}

export type FsType = Pick<typeof fileSystem, 'writeFile' | 'access' | 'mkdir'>
export type NowType = () => number

export class ProtocolLogger {
  private now: NowType
  private fs: FsType
  private logs: LogEntry[]
  private logFileName?: string

  constructor(now: NowType = () => Date.now(), fs: FsType = fileSystem) {
    this.now = now
    this.fs = fs
    this.logs = []
    this.logFileName = undefined
  }

  setLogFileName(logFileName?: string) {
    this.logFileName = logFileName
  }

  getLogFileName(): string | undefined {
    return this.logFileName
  }

  private log(type: LogEntryType, level: LogLevel, data: any) {
    this.logs.push({
      timestamp: this.now(),
      level,
      type,
      data,
    })
  }

  private logCustom(type: CustomLogEntryType, data: any) {
    this.log('CUSTOM', 'DEBUG', {
      type,
      data,
    })
  }

  logProtocolReadStart() {
    this.log('PROTOCOL_READ_START', 'INFO', undefined)
  }

  logProtocolReadEnd() {
    this.log('PROTOCOL_READ_END', 'INFO', undefined)
  }

  logTestRunStart() {
    this.log('TEST_RUN_START', 'INFO', undefined)
  }

  logTestRunEnd() {
    this.log('TEST_RUN_END', 'INFO', undefined)
  }

  logDiscoveredProtocolEnvVars(envVars: { [key: string]: string }) {
    this.log('DISCOVERED_PROTOCOL_ENV_VARS', 'DEBUG', envVars)
  }

  logProtocolVersion(version: string) {
    this.log('PROTOCOL_VERSION', 'DEBUG', version)
  }

  logError(message: string) {
    this.log('MESSAGE', 'ERROR', message)
  }

  logAdapterLoadStart() {
    this.logCustom('ADAPTER_LOAD_START', undefined)
  }

  logAdapterLoadEnd() {
    this.logCustom('ADAPTER_LOAD_END', undefined)
  }

  logAdapterPath(framework: string) {
    this.logCustom('ADAPTER_PATH', framework)
  }

  async write(): Promise<string> {
    if (!this.logFileName) {
      return ''
    }

    const logFileDir = path.dirname(this.logFileName)
    try {
      await this.fs.access(logFileDir)
    } catch (e) {
      await this.fs.mkdir(logFileDir, { recursive: true })
    }

    const loggingJson = JSON.stringify({ logs: this.logs }, null, 2)

    await this.fs.writeFile(this.logFileName, loggingJson)

    log.info('Wrote logging information to', this.logFileName)

    return loggingJson
  }
}
