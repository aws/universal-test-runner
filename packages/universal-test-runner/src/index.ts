// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export { run } from '../src/run'
export { loadAdapter, builtInAdapters } from '../src/loadAdapter'
export { readProtocol } from '../src/readProtocol'
