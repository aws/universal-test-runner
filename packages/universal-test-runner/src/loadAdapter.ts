// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { log } from './log'
import path from 'path'

import { Adapter } from '@sentinel-internal/universal-test-runner-types'

export const builtInAdapters: { [key: string]: string } = {
  jest: '@sentinel-internal/universal-test-adapter-jest',
  maven: '@sentinel-internal/universal-test-adapter-maven',
  gradle: '@sentinel-internal/universal-test-adapter-gradle',
  pytest: '@sentinel-internal/universal-test-adapter-pytest',
  dotnet: '@sentinel-internal/universal-test-adapter-dotnet',
} as const

export async function loadAdapter(rawAdapterModule: string, cwd: string): Promise<Adapter> {
  const adapterModule = builtInAdapters[rawAdapterModule] || rawAdapterModule

  try {
    const adapterImportPath = adapterModule.startsWith('.')
      ? path.join(cwd, adapterModule)
      : adapterModule
    const adapter = await import(adapterImportPath)
    log.info('Loaded adapter from', adapterModule)
    return adapter
  } catch (e) {
    log.error('Failed to load adapter from', adapterModule)
    throw e
  }
}
