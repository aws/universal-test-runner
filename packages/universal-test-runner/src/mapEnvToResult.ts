// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import log from './log'

export type Environment = { [key: string]: string | undefined }

/*
 * Loads a key from the given env, retrieving the value of the first key found
 * in the list of keys, then applies the mapper function and returns the value.
 * Populates envSubset with the raw key and value read from the environment.
 *
 * e.g. Given that the environment variables NODE_ENV="PROD" and
 * NODE_ENVIRONMENT="BETA" are set:
 *
 * ```
 * mapEnvToResult(process.env, ['NODE_ENV', 'NODE_ENVIRONMENT'], val => val?.toLowerCase())
 * ```
 *
 * returns the string "prod". If NODE_ENV was NOT set, it would return "beta".
 */
function mapEnvToResult<T>(
  env: Environment,
  keys: string[],
  mapper: (input: string | undefined) => T,
  envSubset: Environment,
): T {
  let key, value
  for (const currentKey of keys) {
    ;[key, value] = [currentKey, env[currentKey]]
    if (typeof value !== 'undefined') {
      envSubset[key] = value
      break
    }
  }
  if (typeof value !== 'undefined') {
    log.info(`Found ${key} in environment, reading value`)
  }
  return mapper(value)
}

export default mapEnvToResult
