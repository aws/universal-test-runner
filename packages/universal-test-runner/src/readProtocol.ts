// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { mapEnvToResult, Environment } from './mapEnvToResult'
import { log } from './log'
import { ProtocolEnvVars } from './ProtocolEnvVars'

interface TestCase {
  testName: string
  suiteName?: string
  filepath?: string
}

export interface ProtocolResult {
  version: string
  testsToRun: TestCase[]
  logFileName?: string
  reportFormat?: string
}

export function readProtocol(env: Environment): [ProtocolResult, { [key: string]: string }] {
  const rawValues = {}
  const result = {
    version: mapEnvToResult(env, [ProtocolEnvVars.VERSION], readVersion, rawValues),
    testsToRun: mapEnvToResult(env, [ProtocolEnvVars.TESTS_TO_RUN], readTestsToRun, rawValues),
    logFileName: mapEnvToResult(env, [ProtocolEnvVars.LOG_FILE_NAME], readLogFileName, rawValues),
    reportFormat: mapEnvToResult(env, [ProtocolEnvVars.REPORT_FORMAT], readReportFormat, rawValues),
  }
  return [result, rawValues]
}

function readVersion(input: string | undefined): string {
  const DEFAULT_VERSION = '0.1.0'

  if (!input) {
    log.warn('Protocol version not specified! Defaulting to', DEFAULT_VERSION)
  }

  return input ?? DEFAULT_VERSION
}

function readTestsToRun(input: string | undefined): TestCase[] {
  const TEST_CASE_SEPARATOR = '|'
  const TEST_LOCATION_SEPARATOR = '#'

  return (
    input?.split(TEST_CASE_SEPARATOR).map((testCase) => {
      const [testName, suiteName, filepath] = testCase.split(TEST_LOCATION_SEPARATOR).reverse()
      return {
        testName,
        suiteName: suiteName || undefined,
        filepath: filepath || undefined,
      }
    }) ?? []
  )
}

function readLogFileName(logFileName: string | undefined): string | undefined {
  return logFileName || undefined
}

function readReportFormat(reportFormat: string | undefined): string | undefined {
  return reportFormat || undefined
}
