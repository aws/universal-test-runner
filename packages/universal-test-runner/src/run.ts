// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { log } from './log'
import {
  Adapter,
  AdapterInput,
  AdapterOutput,
} from '@aws/universal-test-runner-types'
import { ProtocolResult } from './readProtocol'
import { ErrorCodes } from '../bin/ErrorCodes'

function mapProtocolResultToAdapterInput(protocolResult: ProtocolResult): AdapterInput {
  return {
    testsToRun: protocolResult.testsToRun,
  }
}

export async function run(
  adapter: Adapter,
  protocolResult: ProtocolResult,
): Promise<AdapterOutput> {
  const adapterInput = mapProtocolResultToAdapterInput(protocolResult)
  try {
    log.info('Calling executeTests on adapter...')
    const adapterOutput = await adapter.executeTests(adapterInput)
    log.info('Finished executing tests.')
    if (adapterOutput.exitCode !== 0) {
      log.error(`Test run failed with exit code ${adapterOutput.exitCode}`)
    }
    return adapterOutput
  } catch (e) {
    log.error('Failed to run tests due to an adapter error.', e)
    return { exitCode: ErrorCodes.ADAPTER_ERROR }
  }
}
