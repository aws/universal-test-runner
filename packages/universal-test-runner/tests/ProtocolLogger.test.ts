// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ProtocolLogger, FsType, NowType } from '../src/ProtocolLogger'

jest.mock('../src/log')

describe('ProtocolLogger', () => {
  let logger: ProtocolLogger
  let mockNow: NowType
  let mockFs: FsType

  beforeEach(() => {
    mockNow = (() => {
      let count = 1
      return () => count++
    })()

    mockFs = {
      writeFile: jest.fn(() => Promise.resolve()),
      mkdir: jest.fn<any, any, any>(),
      access: jest.fn(() => Promise.resolve()),
    }

    logger = new ProtocolLogger(mockNow, mockFs)
  })

  it('writes the log file correctly', async () => {
    logger.setLogFileName('llamas.json')
    logger.logProtocolReadStart()
    logger.logProtocolReadEnd()
    logger.logDiscoveredProtocolEnvVars({ llamas: 'Larger than frogs' })
    logger.logAdapterLoadStart()
    logger.logAdapterPath('some/cool/adapter')
    logger.logAdapterLoadEnd()
    logger.logTestRunStart()
    logger.logTestRunEnd()
    logger.logError('some informative error message')

    const result = await logger.write()

    expect(mockFs.writeFile).toHaveBeenCalledWith('llamas.json', result)

    expect(JSON.parse(result)).toMatchSnapshot()
  })
})
