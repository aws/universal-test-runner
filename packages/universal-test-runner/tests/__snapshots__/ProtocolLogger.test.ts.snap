// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtocolLogger writes the log file correctly 1`] = `
{
  "logs": [
    {
      "level": "INFO",
      "timestamp": 1,
      "type": "PROTOCOL_READ_START",
    },
    {
      "level": "INFO",
      "timestamp": 2,
      "type": "PROTOCOL_READ_END",
    },
    {
      "data": {
        "llamas": "Larger than frogs",
      },
      "level": "DEBUG",
      "timestamp": 3,
      "type": "DISCOVERED_PROTOCOL_ENV_VARS",
    },
    {
      "data": {
        "type": "ADAPTER_LOAD_START",
      },
      "level": "DEBUG",
      "timestamp": 4,
      "type": "CUSTOM",
    },
    {
      "data": {
        "data": "some/cool/adapter",
        "type": "ADAPTER_PATH",
      },
      "level": "DEBUG",
      "timestamp": 5,
      "type": "CUSTOM",
    },
    {
      "data": {
        "type": "ADAPTER_LOAD_END",
      },
      "level": "DEBUG",
      "timestamp": 6,
      "type": "CUSTOM",
    },
    {
      "level": "INFO",
      "timestamp": 7,
      "type": "TEST_RUN_START",
    },
    {
      "level": "INFO",
      "timestamp": 8,
      "type": "TEST_RUN_END",
    },
    {
      "data": "some informative error message",
      "level": "ERROR",
      "timestamp": 9,
      "type": "MESSAGE",
    },
  ],
}
`;
